src/System.Linq/src/System/Linq/Select.cs

```diff
@@ -351,7 +388,17 @@ public List<TResult> ToList()
 
             public int GetCount(bool onlyIfCheap)
             {
-                return _source.Count;
+                int count = _source.Count;
+
+                if (!onlyIfCheap)
```

Comment:
Likewise, your reason for doing this isn't obvious from the code alone, so should be commented on. And likely elsewhere, so I won't call out other cases.